README.md
# request-normalizer
**WIP/Incomplete**

This is a utility library that normalizes URLs and other HTTP request information. It is a fork of Nikolay Panov's [url_normalize](https://github.com/niksite/url-normalize) library, with additional features optimized for HTTP request matching.

# Installation
```sh
pip install request-normalizer
```

# Usage

URL only:
```python
>>> from request_normalizer import normalize_url

>>> normalize_url('http://de.wikipedia.org/wiki/Elf (Begriffsklärung)')
'http://de.wikipedia.org/wiki/Elf%20%28Begriffskl%C3%A4rung%29'

>>> normalize_url('https://EXAMPLE.COM:443/page/?user=%7Ejane&q=%5c')
'https://example.com/page/?q=%5C&user=~jane'
```

Complete request:
```python
>>> from request_normalizer import normalize_request
>>> normalize_request(
...     url='https://EXAMPLE.COM:443/page#fragment',
...     headers={'X-Test': 'test', 'accept': 'text/html', 'content-type=application/json'},
...     body=b'{"key_2": "value_2", "key_1": "value_1"}',
...     )
(
    'https://example.com/page#fragment',
    {
        'Accept': 'text/html',
        'Content-Type': 'application/json',
        'X-Test': 'test',
    },
    b'{"key_1": "value_1", "key_2": "value_2"}',
)
```

# Rules
Here is a summary of rules applied to request information:

## URI
* Encode IDN domains
* Only perform percent-encoding where it is essential
* Lowercase URI scheme and host
* Uppercase urlencoded characters (`%XX`)
* Remove dot-segments from non-relative URI paths (`site.com/path/../to/file`)
* For schemes that define a port, use an empty port if the default is desired
* For schemes that define a default authority, use an empty authority if the default is desired
* For schemes that define an empty path to be equivalent to a path of "/", use "/"
* All portions of the URI must be utf-8 encoded NFC from Unicode strings
* (Optional) Sort query parameters
* (Optional) Remove or redact specific query parameters

## Headers
* Camelcase header names (with known exceptions, e.g. `WWW-Authenticate`)
* Sort headers
* (Optional) Remove or redact specific headers

## Request body
For a request body containing JSON or form data:
* Sort keys/fields
* (Optional) Remove or redact specific keys/fields


HISTORY.md
# History

## 2.0.0
* Forked from [url-normalize](https://github.com/niksite/url-normalize)
* Added additional request normalization utilities from [requests-cache](https://github.com/requests-cache/requests-cache)
* Remove support for `?_escaped_fragment_` (for deprecated google crawler feature) 
* Skip removing `utm_source` param (can be done via ignored_params if needed)

**Compatibility:**
* Remove python 2.7 support
* Add type annotations and PEP-561 compliance
* Add tests for python 3.8-3.12 and pypy 3.9-3.10
* Add more default ports for common protocols

**Bugfixes:**
* Handle case where there is a known port but no scheme
* Handle case where there is a known port but no scheme

**Internal:**
* Add CI/dev tools: pre-commit, ruff, mypy, nox, and GitHub Actions

## 1.x (url-normalize)

* 1.4.3: Added LICENSE file
* 1.4.2: Added an optional param sort_query_params (True by default)
* 1.4.1: Added an optional param default_scheme to the url_normalize ('https' by default)
* 1.4.0: A bit of code refactoring and cleanup
* 1.3.3: Support empty string and double slash urls (//domain.tld)
* 1.3.2: Same code support both Python 3 and Python 2.
* 1.3.1: Python 3 compatibility
* 1.2.1: PEP8, setup.py
* 1.1.2: support for shebang (#!) urls
* 1.1.1: using 'http' schema by default when appropriate
* 1.1.0: added handling of IDN domains
* 1.0.0: code pep8
* 0.1.0: forked from Sam Ruby's [urlnorm.py](http://intertwingly.net/blog/2004/08/04/Urlnorm)


tests/test_url_normalize.py
import pytest

from url_normalize.url_normalize import (
    URL,
    deconstruct_url,
    generic_url_cleanup,
    normalize_host,
    normalize_path,
    normalize_port,
    normalize_query,
    normalize_userinfo,
    provide_url_scheme,
    reconstruct_url,
    requote,
    url_normalize,
)


@pytest.mark.parametrize(
    "url, expected",
    [
        ("/../foo", "/foo"),
        ("/./../foo", "/foo"),
        ("/./foo", "/foo"),
        ("/./foo/.", "/foo/"),
        ("//www.foo.com/", "https://www.foo.com/"),
        ("/foo/../bar", "/bar"),
        ("/foo/./bar", "/foo/bar"),
        ("/foo//", "/foo/"),
        ("/foo///bar//", "/foo/bar/"),
        ("/foo/bar/..", "/foo/"),
        ("/foo/bar/../..", "/"),
        ("/foo/bar/../../../../baz", "/baz"),
        ("/foo/bar/../../../baz", "/baz"),
        ("/foo/bar/../../", "/"),
        ("/foo/bar/../../baz", "/baz"),
        ("/foo/bar/../", "/foo/"),
        ("/foo/bar/../baz", "/foo/baz"),
        ("/foo/bar/.", "/foo/bar/"),
        ("/foo/bar/./", "/foo/bar/"),
        ("HTTP://:@example.com/", "http://example.com/"),
        ("http://:@example.com/", "http://example.com/"),
        ("http://@example.com/", "http://example.com/"),
        ("http://127.0.0.1:80/", "http://127.0.0.1/"),
        ("http://example.com:081/", "http://example.com:81/"),
        ("http://example.com:80/", "http://example.com/"),
        ("http://example.com", "http://example.com/"),
        ("http://example.com/?b&a", "http://example.com/?a&b"),
        ("http://example.com/?q=%5c", "http://example.com/?q=%5C"),
        ("http://example.com/?q=%C7", "http://example.com/?q=%EF%BF%BD"),
        ("http://example.com/?q=C%CC%A7", "http://example.com/?q=%C3%87"),
        ("http://EXAMPLE.COM/", "http://example.com/"),
        ("http://example.com/%7Ejane", "http://example.com/~jane"),
        ("http://example.com/a/../a/b", "http://example.com/a/b"),
        ("http://example.com/a/./b", "http://example.com/a/b"),
        (
            "http://example.com/#!5753509/hello-world",
            "http://example.com/?_escaped_fragment_=5753509/hello-world",
        ),
        (
            "http://USER:pass@www.Example.COM/foo/bar",
            "http://USER:pass@www.example.com/foo/bar",
        ),
        ("http://www.example.com./", "http://www.example.com/"),
        ("http://www.foo.com:80/foo", "http://www.foo.com/foo"),
        ("http://www.foo.com.:81/foo", "http://www.foo.com:81/foo"),
        ("http://www.foo.com./foo/bar.html", "http://www.foo.com/foo/bar.html"),
        ("http://www.foo.com/foo/bar.html/../bar.html", "http://www.foo.com/bar.html"),
        ("http://www.foo.com/%7Ebar", "http://www.foo.com/~bar"),
        ("http://www.foo.com/%7ebar", "http://www.foo.com/~bar"),
        (
            "пример.испытание/Служебная:Search/Test",
            "https://xn--e1afmkfd.xn--80akhbyknj4f/%D0%A1%D0%BB%D1%83%D0%B6%D0%B5%D0%B1%D0%BD%D0%B0%D1%8F:Search/Test",
        ),
    ],
)
def test_url_normalize(url, expected):
    """Test main url_normalize function with a representative set of URLs"""
    assert url_normalize(url) == expected


@pytest.mark.parametrize(
    "url",
    [
        "-",
        "",
        "/..foo",
        "/.foo",
        "/foo..",
        "/foo.",
        "ftp://user:pass@ftp.foo.net/foo/bar",
        "http://127.0.0.1/",
        "http://example.com:8080/",
        "http://example.com/?a&b",
        "http://example.com/?q=%5C",
        "http://example.com/?q=%C3%87",
        "http://example.com/?q=%E2%85%A0",
        "http://example.com/",
        "http://example.com/~jane",
        "http://example.com/a/b",
        "http://example.com/FOO",
        "http://user:password@example.com/",
        "http://www.foo.com:8000/foo",
        # from rfc2396bis
        "ftp://ftp.is.co.za/rfc/rfc1808.txt",
        "http://www.ietf.org/rfc/rfc2396.txt",
        "ldap://[2001:db8::7]/c=GB?objectClass?one",
        "mailto:John.Doe@example.com",
        "news:comp.infosystems.www.servers.unix",
        "tel:+1-816-555-1212",
        "telnet://192.0.2.16:80/",
        "urn:oasis:names:specification:docbook:dtd:xml:4.1.2",
    ],
)
def test_url_normalize__no_change(url):
    """Assert url_normalize does not change URI if not required.

    http://www.intertwingly.net/wiki/pie/PaceCanonicalIds
    """
    assert url_normalize(url) == url


def test_url_normalize__with_http_scheme():
    """Assert we could use http scheme as default."""
    url = "//www.foo.com/"
    expected = "http://www.foo.com/"
    assert url_normalize(url, default_scheme="http") == expected


def test_url_normalize__with_no_params_sorting():
    """Assert we could use http scheme as default."""
    url = "http://www.foo.com/?b=1&a=2"
    expected = "http://www.foo.com/?b=1&a=2"
    assert url_normalize(url, sort_query_params=False) == expected


@pytest.mark.parametrize(
    "url, expected",
    [
        (
            "http://site.com",
            URL(
                fragment="",
                host="site.com",
                path="",
                port="",
                query="",
                scheme="http",
                userinfo="",
            ),
        ),
        (
            "http://user@www.example.com:8080/path/index.html?param=val#fragment",
            URL(
                fragment="fragment",
                host="www.example.com",
                path="/path/index.html",
                port="8080",
                query="param=val",
                scheme="http",
                userinfo="user@",
            ),
        ),
    ],
)
def test_deconstruct_url(url, expected):
    assert deconstruct_url(url) == expected


@pytest.mark.parametrize(
    "url, expected",
    [
        (
            URL(
                fragment="",
                host="site.com",
                path="",
                port="",
                query="",
                scheme="http",
                userinfo="",
            ),
            "http://site.com",
        ),
        (
            URL(
                fragment="fragment",
                host="www.example.com",
                path="/path/index.html",
                port="8080",
                query="param=val",
                scheme="http",
                userinfo="user@",
            ),
            "http://user@www.example.com:8080/path/index.html?param=val#fragment",
        ),
    ],
)
def test_reconstruct_url(url, expected):
    assert reconstruct_url(url) == expected


@pytest.mark.parametrize(
    "url, expected",
    [
        ("//site/#!fragment", "//site/?_escaped_fragment_=fragment"),
        ("//site/?utm_source=some source&param=value", "//site/?param=value"),
        ("//site/?utm_source=some source", "//site/"),
        ("//site/?param=value&utm_source=some source", "//site/?param=value"),
        ("//site/page", "//site/page"),
        ("//site/?& ", "//site/"),
    ],
)
def test_generic_url_cleanup(url, expected):
    assert generic_url_cleanup(url) == expected


@pytest.mark.parametrize(
    "fragment, expected",
    [
        ("", ""),
        ("fragment", "fragment"),
        ("пример", "%D0%BF%D1%80%D0%B8%D0%BC%D0%B5%D1%80"),
        ("!fragment", "%21fragment"),
        ("~fragment", "~fragment"),
    ],
)
def test_normalize_fragment(fragment, expected):
    assert requote(fragment, safe="~") == expected


@pytest.mark.parametrize(
    "host, expected",
    [
        ("site.com", "site.com"),
        ("SITE.COM", "site.com"),
        ("site.com.", "site.com"),
        ("пример.испытание", "xn--e1afmkfd.xn--80akhbyknj4f"),
    ],
)
def test_normalize_host(host, expected):
    assert normalize_host(host) == expected


@pytest.mark.parametrize(
    "path, expected",
    [
        ("..", "/"),
        ("", "/"),
        ("/../foo", "/foo"),
        ("/..foo", "/..foo"),
        ("/./../foo", "/foo"),
        ("/./foo", "/foo"),
        ("/./foo/.", "/foo/"),
        ("/.foo", "/.foo"),
        ("/", "/"),
        ("/foo..", "/foo.."),
        ("/foo.", "/foo."),
        ("/FOO", "/FOO"),
        ("/foo/../bar", "/bar"),
        ("/foo/./bar", "/foo/bar"),
        ("/foo//", "/foo/"),
        ("/foo///bar//", "/foo/bar/"),
        ("/foo/bar/..", "/foo/"),
        ("/foo/bar/../..", "/"),
        ("/foo/bar/../../../../baz", "/baz"),
        ("/foo/bar/../../../baz", "/baz"),
        ("/foo/bar/../../", "/"),
        ("/foo/bar/../../baz", "/baz"),
        ("/foo/bar/../", "/foo/"),
        ("/foo/bar/../baz", "/foo/baz"),
        ("/foo/bar/.", "/foo/bar/"),
        ("/foo/bar/./", "/foo/bar/"),
    ],
)
def test_normalize_path(path, expected):
    assert normalize_path(path, "http") == expected


@pytest.mark.parametrize(
    "port, expected",
    [
        ("8080", "8080"),
        ("", ""),
        ("80", ""),
        ("string", "string"),
    ],
)
def test_normalize_port(port, expected):
    assert normalize_port(port, 'http') == expected


@pytest.mark.parametrize(
    "url, expected",
    [
        ("", ""),
        ("param1=val1&param2=val2", "param1=val1&param2=val2"),
        ("Ç=Ç", "%C3%87=%C3%87"),
        ("%C3%87=%C3%87", "%C3%87=%C3%87"),
        ("q=C%CC%A7", "q=%C3%87"),
    ],
)
def test_normalize_queryd(url, expected):
    assert normalize_query(url) == expected


@pytest.mark.parametrize(
    "url, expected",
    [
        (":@", ""),
        ("", ""),
        ("@", ""),
        ("user:password@", "user:password@"),
        ("user@", "user@"),
    ],
)
def test_normalize_userinfo(url, expected):
    assert normalize_userinfo(url) == expected


@pytest.mark.parametrize(
    "url, expected",
    [
        ("", ""),
        ("-", "-"),
        ("/file/path", "/file/path"),
        ("//site/path", "https://site/path"),
        ("ftp://site/", "ftp://site/"),
        ("site/page", "https://site/page"),
    ],
)
def test_provide_url_scheme_result_is_expected(url, expected):
    """Assert we got expected results from the provide_url_scheme function."""
    assert provide_url_scheme(url) == expected


def test_provide_url_scheme_accept_default_scheme_param():
    """Assert we could provide default_scheme param other than https."""
    url = "//site/path"
    expected = "http://site/path"
    assert provide_url_scheme(url, default_scheme="http") == expected


tests/test_normalize_host.py
"""Tests for normalize_host function."""
import pytest

from url_normalize.url_normalize import normalize_host

EXPECTED_DATA = {
    "site.com": "site.com",
    "SITE.COM": "site.com",
    "site.com.": "site.com",
    "пример.испытание": "xn--e1afmkfd.xn--80akhbyknj4f",
}


@pytest.mark.parametrize("host, expected", EXPECTED_DATA.items())
def test_normalize_host_result_is_expected(host, expected):
    """Assert we got expected results from the normalize_host function."""
    assert normalize_host(host) == expected


tests/test_normalize_path.py
"""Tests for normalize_path function."""
import pytest
from url_normalize.url_normalize import normalize_path

EXPECTED_DATA = {
    "..": "/",
    "": "/",
    "/../foo": "/foo",
    "/..foo": "/..foo",
    "/./../foo": "/foo",
    "/./foo": "/foo",
    "/./foo/.": "/foo/",
    "/.foo": "/.foo",
    "/": "/",
    "/foo..": "/foo..",
    "/foo.": "/foo.",
    "/FOO": "/FOO",
    "/foo/../bar": "/bar",
    "/foo/./bar": "/foo/bar",
    "/foo//": "/foo/",
    "/foo///bar//": "/foo/bar/",
    "/foo/bar/..": "/foo/",
    "/foo/bar/../..": "/",
    "/foo/bar/../../../../baz": "/baz",
    "/foo/bar/../../../baz": "/baz",
    "/foo/bar/../../": "/",
    "/foo/bar/../../baz": "/baz",
    "/foo/bar/../": "/foo/",
    "/foo/bar/../baz": "/foo/baz",
    "/foo/bar/.": "/foo/bar/",
    "/foo/bar/./": "/foo/bar/",
}


@pytest.mark.parametrize("path, expected", EXPECTED_DATA.items())
def test_normalize_host_result_is_expected(path, expected):
    """Assert we got expected results from the normalize_path function."""
    assert normalize_path(path, "http") == expected


noxfile.py
from os import getenv

import nox
from nox_poetry import session

nox.options.reuse_existing_virtualenvs = True
nox.options.sessions = ['lint', 'cov']

PYTHON_VERSIONS = ['3.7', '3.8', '3.9', '3.10', '3.11', '3.12', 'pypy3.9']
DEFAULT_COVERAGE_FORMATS = ['html', 'term']


@session(python=PYTHON_VERSIONS)
def test(session):
    """Run tests in a separate virtualenv per python version"""
    session.install('.', 'pytest', 'pytest-socket')
    session.run('pytest', '-rs')


@session(python=False, name='cov')
def coverage(session):
    """Run tests and generate coverage report (in current virtualenv)"""
    cmd = ['pytest', '-rs', '--cov']

    # Add coverage formats
    cov_formats = session.posargs or DEFAULT_COVERAGE_FORMATS
    cmd += [f'--cov-report={f}' for f in cov_formats]

    # Add verbose flag, if set by environment
    if getenv('PYTEST_VERBOSE'):
        cmd += ['--verbose']
    session.run(*cmd)


@session(python=False)
def lint(session):
    """Run linters and code formatters via pre-commit"""
    session.run('pre-commit', 'run', '--all-files')


url_normalize/tools.py
"""Url normalize tools"""
import re
import unicodedata
from collections import namedtuple
from urllib.parse import unquote as unquote_orig
from urllib.parse import urlsplit, urlunsplit

URL = namedtuple(
    "URL", ["scheme", "userinfo", "host", "port", "path", "query", "fragment"]
)


def deconstruct_url(url):
    """Tranform the url into URL structure.

    Params:
        url : string : the URL

    Returns:
        URL

    """
    scheme, auth, path, query, fragment = urlsplit(url.strip())
    (userinfo, host, port) = re.search("([^@]*@)?([^:]*):?(.*)", auth).groups()
    return URL(
        fragment=fragment,
        host=host,
        path=path,
        port=port,
        query=query,
        scheme=scheme,
        userinfo=userinfo or "",
    )


def reconstruct_url(url):
    """Reconstruct string url from URL.

    Params:
        url : URL object instance

    Returns:
        string : reconstructed url string

    """
    auth = (url.userinfo or "") + url.host
    if url.port:
        auth += ":" + url.port
    return urlunsplit((url.scheme, auth, url.path, url.query, url.fragment))


def unquote(string, charset="utf-8"):
    """Unquote and normalize unicode string.

    Params:
        string : string to be unquoted
        charset : string : optional : output encoding

    Returns:
        string : an unquoted and normalized string

    """
    string = unquote_orig(string)
    string = unicodedata.normalize("NFC", string).encode(charset)
    return string


url_normalize/url_normalize.py
"""URL normalize main module."""
import re
import unicodedata
from collections import namedtuple
from urllib.parse import quote, unquote, urlsplit, urlunsplit

import idna

URL = namedtuple("URL", ["scheme", "userinfo", "host", "port", "path", "query", "fragment"])

DEFAULT_PORT = {
    "ftp": "21",
    "gopher": "70",
    "http": "80",
    "https": "443",
    "news": "119",
    "nntp": "119",
    "snews": "563",
    "snntp": "563",
    "telnet": "23",
    "ws": "80",
    "wss": "443",
}
DEFAULT_CHARSET = "utf-8"
DEFAULT_SCHEME = "https"


def url_normalize(
    url, charset=DEFAULT_CHARSET, default_scheme=DEFAULT_SCHEME, sort_query_params=True
):
    """URI normalization routine.

    Sometimes you get an URL by a user that just isn't a real
    URL because it contains unsafe characters like ' ' and so on.
    This function can fix some of the problems in a similar way
    browsers handle data entered by the user:

    >>> url_normalize('http://de.wikipedia.org/wiki/Elf (Begriffsklärung)')
    'http://de.wikipedia.org/wiki/Elf%20%28Begriffskl%C3%A4rung%29'

    Params:
        charset : string : optional
            The target charset for the URL if the url was given as unicode string.

    Returns:
        string : a normalized url

    """
    if not url:
        return url
    url = provide_url_scheme(url, default_scheme)
    url = generic_url_cleanup(url)
    url_elements = deconstruct_url(url)
    url_elements = url_elements._replace(
        scheme=url_elements.scheme.lower(),
        userinfo=normalize_userinfo(url_elements.userinfo),
        host=normalize_host(url_elements.host, charset),
        query=normalize_query(url_elements.query, sort_query_params),
        fragment=requote(url_elements.fragment, safe="~"),
    )
    url_elements = url_elements._replace(
        port=normalize_port(url_elements.port, url_elements.scheme),
        path=normalize_path(url_elements.path, url_elements.scheme),
    )
    url = reconstruct_url(url_elements)
    return url


def provide_url_scheme(url, default_scheme=DEFAULT_SCHEME):
    """Make sure we have valid url scheme.

    Params:
        url : string : the URL
        default_scheme : string : default scheme to use, e.g. 'https'

    Returns:
        string : updated url with validated/attached scheme
    """
    has_scheme = ":" in url[:7]
    is_universal_scheme = url.startswith("//")
    is_file_path = url == "-" or (url.startswith("/") and not is_universal_scheme)
    if not url or has_scheme or is_file_path:
        return url
    if is_universal_scheme:
        return default_scheme + ":" + url
    return default_scheme + "://" + url


def generic_url_cleanup(url):
    """Cleanup the URL from unnecessary data and convert to final form.

    Converts shebang urls to final form, removed unnecessary data from the url.

    Params:
        url : string : the URL

    Returns:
        string : update url
    """
    url = url.replace("#!", "?_escaped_fragment_=")
    url = re.sub(r"utm_source=[^&]+&?", "", url)
    url = url.rstrip("&? ")
    return url


def normalize_userinfo(userinfo):
    """Normalize userinfo part of the url.

    Params:
        userinfo : string : url userinfo, e.g., 'user@'

    Returns:
        string : normalized userinfo data.
    """
    if userinfo in ["@", ":@"]:
        return ""
    return userinfo


def normalize_host(host, charset=DEFAULT_CHARSET):
    """Normalize host part of the url.

    Lowercase and strip of final dot.
    Also, take care about IDN domains.

    Params:
        host : string : url host, e.g., 'site.com'

    Returns:
        string : normalized host data.
    """
    host = host.lower()
    host = host.strip(".")
    # Skip IDN normalization for URIs that do not contain a domain name
    if '.' not in host:
        return host
    return idna.encode(host, uts46=True, transitional=True).decode(charset)


def normalize_port(port, scheme):
    """Normalize port part of the url.

    Remove mention of default port number

    Params:
        port : string : url port, e.g., '8080'
        scheme : string : url scheme, e.g., 'http'

    Returns:
        string : normalized port data.
    """
    if not port.isdigit():
        return port
    port = str(int(port))
    if DEFAULT_PORT[scheme] == port:
        return ""
    return port


def normalize_path(path, scheme):
    """Normalize path part of the url.

    Remove mention of default path number

    Params:
        path : string : url path, e.g., '/section/page.html'
        scheme : string : url scheme, e.g., 'http'

    Returns:
        string : normalized path data.
    """
    # Only perform percent-encoding where it is essential.
    # Always use uppercase A-through-F characters when percent-encoding.
    # All portions of the URI must be utf-8 encoded NFC from Unicode strings
    path = requote(path, safe="~:/?#[]@!$&'()*+,;=")
    # Prevent dot-segments appearing in non-relative URI paths.
    if scheme in ["", "http", "https", "ftp", "file"]:
        output, part = [], None
        parts = path.split("/")
        last_idx = len(parts) - 1
        for idx, part in enumerate(parts):
            if part == "":
                if not output:
                    output.append(part)
            elif part == ".":
                pass
            elif part == "..":
                if len(output) > 1:
                    output.pop()
            elif re.search(r'\.', part) and idx != last_idx:
                pass
            else:
                output.append(part)
        if part in ["", ".", ".."]:
            output.append("")
        path = "/".join(output)
    # For schemes that define an empty path to be equivalent to a path of "/",
    # use "/".
    if not path and scheme in ["http", "https", "ftp", "file"]:
        path = "/"
    return path


def normalize_query(query, sort_query_params=True):
    """Normalize query part of the url.

    Params:
        query : string : url query, e.g., 'param1=val1&param2=val2'

    Returns:
        string : normalized query data.
    """
    param_arr = [
        "=".join([requote(t, safe="~:/?#[]@!$'()*+,;=") for t in q.split("=", 1)])
        for q in query.split("&")
    ]
    if sort_query_params:
        param_arr = sorted(param_arr)
    query = "&".join(param_arr)
    return query


def deconstruct_url(url):
    """Tranform the url into URL structure.

    Params:
        url : string : the URL

    Returns:
        URL
    """
    scheme, auth, path, query, fragment = urlsplit(url.strip())
    (userinfo, host, port) = re.search("([^@]*@)?([^:]*):?(.*)", auth).groups()
    return URL(
        fragment=fragment,
        host=host,
        path=path,
        port=port,
        query=query,
        scheme=scheme,
        userinfo=userinfo or "",
    )


def reconstruct_url(url):
    """Reconstruct string url from URL.

    Params:
        url : URL object instance

    Returns:
        string : reconstructed url string
    """
    auth = (url.userinfo or "") + url.host
    if url.port:
        auth += ":" + url.port
    return urlunsplit((url.scheme, auth, url.path, url.query, url.fragment))


def requote(string, charset="utf-8", safe="/"):
    """Unquote and requote unicode string to normalize.

    Params:
        string : string to be unquoted
        charset : string : optional : output encoding

    Returns:
        string : an unquoted and normalized string
    """
    string = unquote(string)
    string = unicodedata.normalize("NFC", string).encode(charset)
    return quote(string, safe)
